Mathlib/Order/LawfulPreorder.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2017 Pierre Quinton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre Quinton
+-/
+import Mathlib.Order.SetNotation
+import Mathlib.Order.Bounds.Defs
+import Mathlib.Order.Defs.PartialOrder
+
+/-!
+# Lawful Suprema and Infima
+
+This file introduces type classes that enforce correctness properties for suprema and infima
+operations on preorders, ensuring they return actual least upper bounds and greatest lower bounds
+whenever they exist.
+
+## Main declarations
+
+
+- `LawfulSupPreorder`: Typeclass ensuring that whenever a set `s` has a least upper bound, `sSup s`
+  returns a least upper bound.
+- `LawfulInfPreorder`: Typeclass ensuring that whenever a set `s` has a greatest lower bound,
+  `sInf s` returns a greatest lower bound.
+- `LawfulPreorder`: Typeclass combining both lawful suprema and infima properties.
```

Comment:
I don't like this name, it says nothing about suprema and infima. If you go with my `LawfulSup` suggestion, you might name this `LawfulSupInf` (for lack of a better suggestion).